onnxruntime/core/providers/qnn/builder/opbuilder/instance_norm_op_builder.cc

```diff
@@ -0,0 +1,136 @@
+// The nodes from 2nd call of GetCapability get their layout transformed to NHWC.
+// Therefore, we need to check the node domain to determine if the layout has been transformed.
+Status InstanceNormOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
+                                            const NodeUnit& node_unit,
+                                            const logging::Logger& logger,
+                                            bool is_quantized_model) const {
+  ORT_UNUSED_PARAMETER(logger);
+
+  const auto float_elem_type = ONNX_NAMESPACE::Utils::DataTypeUtils::ToType("float");
+
+  // Check input type is float for CPU.
+  const auto& inputs = node_unit.Inputs();
+  ONNX_NAMESPACE::DataType input_data_type = inputs[0].node_arg.Type();
+  if (!is_quantized_model && input_data_type != float_elem_type) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm data type " + *input_data_type + " is not supported in CPU backend.");
+  }
+
+  // Also check output type is float for CPU.
+  const auto& outputs = node_unit.Outputs();
+  ONNX_NAMESPACE::DataType output_data_type = outputs[0].node_arg.Type();
+  if (!is_quantized_model && output_data_type != float_elem_type) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm data type " + *output_data_type + " is not supported in CPU backend.");
+  }
+
+  std::vector<uint32_t> input_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape of input 0");
+  const size_t input_rank = input_shape.size();
+
+  if (input_rank <= 2 || input_rank > 4) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm only supports input ranks of size 3 or 4.");
+  }
+
+  if (node_unit.Domain() != kMSInternalNHWCDomain) {  // Layout has not been transformed to NHWC, so do it here.
+    const uint32_t chans = input_shape[1];
+    for (size_t i = 1; i < input_rank - 1; ++i) {
+      input_shape[i] = input_shape[i + 1];
+    }
+
+    input_shape[input_rank - 1] = chans;
+  }
+
+  const uint32_t num_channels = input_shape.back();
+
+  std::vector<uint32_t> scale_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[1].node_arg, scale_shape), "Cannot get shape of input 1 (scale)");
+  if (scale_shape.size() != 1 || scale_shape[0] != num_channels) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm input 1 (scale) must have 1D shape [channel].");
+  }
+
+  std::vector<uint32_t> bias_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[2].node_arg, bias_shape), "Cannot get shape of input 2 (bias)");
+  if (bias_shape.size() != 1 || bias_shape[0] != num_channels) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "QNN InstanceNorm input 2 (bias) must have 1D shape [channel].");
+  }
+
+  NodeAttrHelper node_helper(node_unit);
+  const float epsilon = node_helper.Get("epsilon", 0.0f);
```

Comment:
epsilon default value is 1e-05f according to Onnx spec.

Reply:
Yes, the default value is correctly specified in `ProcessAttributesAndOutputs`. We don't actually need to do any validation of epsilon here. I'll remove this code.

Reply:
re-added this block of code. It's also now using 1e-05f